### PR TITLE
chore(atomix): rename and restructure raft entry types

### PR DESCRIPTION
**Pull request by deepthidevaki**
## Description

This PR is in preparation to moving Raft entries to SBE serialization. #6310 . It contains mostly renaming and re-organizing classes. There is no logic change. 
* Introduce new structure for entry classes. `RaftLogEntry` is the raft entry that will be appended. It encapsulates a term and an entry which can be either an ApplicationEntry, ConfigurationEntry or InitialEntry. 
* Rename ZeebeEntry to ApplicationEntry
* Renamed Indexed<> to `IndexedRaftRecord` and removed templating.

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...develop?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/0.25`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

### CODE REVIEW

File: atomix/cluster/src/main/java/io/atomix/raft/storage/log/entry/RaftLogEntry.java [hunk 1]
```diff
@@ -21,12 +21,42 @@
...
+  public ConfigurationEntry getConfigurationEntry() {
+    return (ConfigurationEntry) entry;
+  }
+
+  public boolean isInitialEntry() {
+    return entry instanceof InitializeEntry;
+  }
+
+  public InitializeEntry getInitialEntry() {
+    return (InitializeEntry) entry;
```
Comment (MiguelPires) [inline atomix/cluster/src/main/java/io/atomix/raft/storage/log/entry/RaftLogEntry.java:59]: These never get called right?

File: atomix/cluster/src/main/java/io/atomix/raft/storage/log/entry/InitializeEntry.java [hunk 2]
```diff
@@ -25,9 +25,4 @@
  * following leader changes. Initialize entries are always the first entry to be committed at the
  * start of a leader's term.
  */
-public class InitializeEntry extends TimestampedEntry {
-
-  public InitializeEntry(final long term, final long timestamp) {
-    super(term, timestamp);
-  }
-}
+public class InitializeEntry implements RaftEntry {}
```
Comment (MiguelPires) [inline atomix/cluster/src/main/java/io/atomix/raft/storage/log/entry/InitializeEntry.java:28]: Is there any reason why we're removing the timestamp for this but keeping it in the ConfigurationEntry?

Comment (MiguelPires) [reply atomix/cluster/src/main/java/io/atomix/raft/storage/log/entry/InitializeEntry.java:28]: Can we rename this to `InitialEntry`? We alternate between the two names throughout the code so we can take this opportunity to make it consistent

**Review by miguelpires — COMMENTED** [review]
Just a few minor questions and comments.
It seems a bit weird to have an empty interface (RaftEntry) with an empty implementation (InitializeEntry). Plus the `(is/get)InitialEntry` methods in the RaftLogEntry are never used. It looks like the abstraction doesn't fit our usage perfectly but the alternatives also don't seem better

**Comment by deepthidevaki** [comment]
> Plus the `(is/get)InitialEntry` methods in the RaftLogEntry are never used. 

These will be used when moving to SBE.

> It looks like the abstraction doesn't fit our usage perfectly but the alternatives also don't seem better

The alternative is stick with the current structure. RaftLogEntry is an abstract class with field term. The other types extends RaftLogEntry. What do you think?

File: atomix/cluster/src/main/java/io/atomix/raft/storage/log/entry/InitializeEntry.java (hunk 2 above)
Comment (deepthidevaki) [reply atomix/cluster/src/main/java/io/atomix/raft/storage/log/entry/InitializeEntry.java:28]: > Is there any reason why we're removing the timestamp for this but keeping it in the ConfigurationEntry?

Timestamp in ConfigurationEntry is used elsewhere. Timestamp for the other entries are not required.

**Review by miguelpires — APPROVED** [review]
I think this is fine for now. The rest looks good :+1: Sorry for the delay in reviewing this

**Pull request merged**